Added http-api with authenticator

Jess3Jane opened the pull request:
- Added http-api
  - Added authenticator middeware to http-api
  - Removed old qaul-http api
  - Changed identity to be Copy/Clone
  - Added user stub
  - Cleaned up a few imports

Jess3Jane committed:
Added http-api with authenticator

  - Added http-api
  - Added authenticator middeware to http-api
  - Removed old qaul-http api
  - Changed identity to be Copy/Clone
  - Added user stub
  - Cleaned up a few imports

spacekookie commented on Cargo.toml:
```diff
@@ -11,6 +11,9 @@ members = [
     "libqaul/service/core",
 
+	# qaul.net http-api
+	"libqaul/http-api",
```
Can you please use spaces for indentation

spacekookie commented on libqaul/http-api/Cargo.toml:
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "http-api"
```
The crate name should probably be more descriptive like `qaul-http` :sweat_smile:

spacekookie commented on libqaul/http-api/src/lib.rs:
```diff
@@ -0,0 +1,78 @@
+
+impl QaulCore {
+    pub fn new(qaul: Arc<Qaul>) -> QaulCore {
+        QaulCore{
```
Formatting is a bit weird here. Also maybe just use `Self`

Jess3Jane replied:
I changed things to use `Self` and reformatted a bit, if you have any other formatting concerns let me know

spacekookie reviewed (approved):
There's some small formatting changes, but otherwise this looks very good.

I assume that part of what the authenticator does here will be able to be moved to the `auth` module in `libqaul` at some point, but having it here for now is totally cool to get started

Jess3Jane committed:
Cleaned up formatting

Jess3Jane committed:
Changed functions to use `Self`

Jess3Jane committed:
Give http-api a more descriptive name

Jess3Jane committed:
Derive `Clone` on `Qaul`

So for the moment `Qaul` will have everything interiorly mutable, which
makes sense but also means we can derive `Clone` for quite cheap. This
enables us to pass around references to the main `Qaul` instance but
make them owned pointers to the same data when we need to: such as when
sharing them between threads

Jess3Jane commented:
I also added a commit that derives `Clone` on `Qaul` to allow references to be shared across thread boundaries as discussed on the Zulip

I'm still not overly sold on this way of doing it but we need some way to share `&Qaul` across thread boundaries

spacekookie commented:
Yea, agreed. It feels like more of a hack for now. What we could possibly do is just do a type alias for `Qaul = Arc<ActualQaul>` or something. But we can bikeshed this later too.
